src/Groups/group_constructors.jl

```diff
@@ -844,38 +846,83 @@ julia> relators(g)
 
-function quaternion_group(::Type{T}, n::IntegerUnion) where T <: GAPGroup
-  # FIXME: resolve naming: dicyclic vs (generalized) quaternion: only the
-  # former should be for any n divisible by 4; the latter only for powers of 2.
-  # see also debate on the GAP side (https://github.com/gap-system/gap/issues/2725)
+quaternion_group(::Type{T}, n::IntegerUnion) where {T<:Union{GAPGroup,PcGroup,SubPcGroup}} =
+  dicyclic_group(T, n)
+
+@doc raw"""
+    is_quaternion_group(G::GAPGroup)
+
+Return `true` if `G` is isomorphic to a (generalized) quaternion group
+of order $2^{k+1}, k \geq 2$, and `false` otherwise.
+
+# Examples
+```jldoctest
+julia> is_quaternion_group(small_group(8, 3))
+false
+
+julia> is_quaternion_group(transitive_group(8, 5))
+true
+```
+"""
+@gapattribute is_quaternion_group(G::GAPGroup) =
+  GAP.Globals.IsGeneralisedQuaternionGroup(GapObj(G))::Bool
+
+"""
+    dicyclic_group(::Type{T} = PcGroup, n::IntegerUnion)
+
+Return the dicyclic group of order `n`, as an instance of `T`,
+where `n` is a multiple of 4 and `T` is in
+{`PcGroup`, `SubPcGroup`, `PermGroup`,`FPGroup`, `SubFPGroup`}.
+
+# Examples
+```jldoctest
+julia> g = dicyclic_group(12)
+Pc group of order 12
+
+julia> dicyclic_group(PermGroup, 12)
+Permutation group of degree 12
+
+julia> g = dicyclic_group(FPGroup, 12)
+Finitely presented group of order 12
+
+julia> relators(g)
+3-element Vector{FPGroupElem}:
+ r^2*s^-3
+ s^6
+ r^-1*s*r*s
+```
+"""
+dicyclic_group(n::IntegerUnion) = dicyclic_group(PcGroup, n)
+
+function dicyclic_group(::Type{T}, n::IntegerUnion) where {T<:GAPGroup}
   @assert iszero(mod(n, 4))
-  return T(GAP.Globals.QuaternionGroup(_gap_filter(T), n)::GapObj)
+  return T(GAP.Globals.DicyclicGroup(_gap_filter(T), n)::GapObj)
 end
 
 # Delegating to the GAP constructor via `_gap_filter` does not work here.
-function quaternion_group(::Type{T}, n::IntegerUnion) where T <: Union{PcGroup, SubPcGroup}
+function dicyclic_group(::Type{T}, n::IntegerUnion) where {T<:Union{PcGroup,SubPcGroup}}
   @assert iszero(mod(n, 4))
-  return T(GAP.Globals.QuaternionGroup(GAP.Globals.IsPcGroup, n)::GapObj)
+  return T(GAP.Globals.DicyclicGroup(GAP.Globals.IsPcGroup, n)::GapObj)
 end
 
 @doc raw"""
-    is_quaternion_group(G::GAPGroup)
+    is_dicyclic_group(G::GAPGroup)
 
-Return `true` if `G` is isomorphic to a (generalized) quaternion group
-of order $2^{k+1}, k \geq 2$, and `false` otherwise.
+Return `true` if `G` is isomorphic to a dicyclic group
+of order $4n, n > 1$, and `false` otherwise.
 
 # Examples
 ```jldoctest
-julia> is_quaternion_group(small_group(8, 3))
+julia> is_dicyclic_group(small_group(8, 3))
 false
 
-julia> is_quaternion_group(small_group(8, 4))
+julia> is_dicyclic_group(small_group(8, 4))
 true
 ```
 """
-@gapattribute is_quaternion_group(G::GAPGroup) = GAP.Globals.IsQuaternionGroup(GapObj(G))::Bool
+@gapattribute is_dicyclic_group(G::GAPGroup) =
+  GAP.Globals.IsQuaternionGroup(GapObj(G))::Bool
```

Comment:
I am getting confused.
GAP does the following:
```
gap> G:= DicyclicGroup(24);;
gap> IsQuaternionGroup(G);
false
gap> G:= DicyclicGroup(32);;
gap> IsQuaternionGroup(G);
true
```
That is, both `QuaternionGroup` and `DicyclicGroup` can be used to create the groups in question,
but `IsQuaternionGroup` regards only the dicyclic groups of 2-power order as (generalized) quaternion groups.
In fact, GAP seems to have no function that checks whether a group is dicyclic.

Reply:
As far as I see, the GAP function `DoComputeGeneralisedQuaternionGenerators` (which is called by `IsQuaternionGroup`) would work as an `is_dicyclic` check if one just omits the check for 2-power order.